File: typescript/src/lib/data-loader.ts
import {
  ALARM_START_DATE,
  EQUIPMENT_CODE_PATTERN,
  PATHS,
  TARGET_SITES,
} from './config'
import { readCsvFile } from './csv'

export type WorkOrder = {
  WO_NO: string
  CUSTOMER_NO: number
  CUSTOMER_SITE_NO: number
  customer_site_name: string
  CONTRACT: string
  WORK_ORDER_TYPE_ENG: string
  WORK_TYPE_DESCRIPTION_ENG: string
  WORK_DESCRIPTION: string
  WORK_ORDER_PERFORMED_ACTION: string
  PM_NO: string
  PM_ACTION_DESCR: string
  WORK_STARTED_DATETIME: string
  WORK_FINISHED_DATETIME: string
  WORKTIME_HOURS: string
  WORK_FINISHED_DAYS: string
  started: Date | null
  finished: Date | null
  hours: number
  days: number
  equipmentCodes: Set<string>
}

export type Alarm = {
  ALERT_ID: string
  ALERT_EVENT_ID: string
  CUSTOMER_NO: number
  CUSTOMER_SITE_NO: number
  CUSTOMER_SITE_NAME: string
  EVENT_TIME: Date | null
  EVENT_DESCRIPTION: string
  ALERT_TYPE: string
  LOG_ONLY: string
  WORKORDER_NO: string
  isActionable: boolean
  equipmentCodes: Set<string>
}

export type MaintenancePlan = {
  PM_NO: number
  CUSTOMER_SITE_NO: number
  customer_site_name: string
  ACTION_DESCR: string
  ACTION_DESCR_ENG: string
  INTERVAL: number
  PM_INTERVAL_UNIT: string
  PLAN_HRS: number
  START_DATE: Date | null
  OBJSTATE: string
  LAST_UPDATED: string
}

export function extractEquipmentCodes(text: string | undefined | null): Set<string> {
  if (!text) return new Set()
  const matches = text.match(EQUIPMENT_CODE_PATTERN) ?? []
  return new Set(matches.map((m) => m.toUpperCase().split('.')[0]))
}

export function parseFiDate(value: string | undefined): Date | null {
  if (!value?.trim()) return null
  const match = value.trim().match(/^(\d{1,2})\.(\d{1,2})\.(\d{4})\s+(\d{1,2}):(\d{2})$/)
  if (!match) return null
  const [, d, m, y, h, min] = match
  return new Date(Number(y), Number(m) - 1, Number(d), Number(h), Number(min))
}

export function parseIsoDate(value: string | undefined): Date | null {
  if (!value?.trim()) return null
  const d = new Date(value.trim().slice(0, 19).replace(' ', 'T'))
  return Number.isNaN(d.getTime()) ? null : d
}

export function addDays(date: Date, days: number): Date {
  const d = new Date(date)
  d.setDate(d.getDate() + days)
  return d
}

export function daysBetween(a: Date, b: Date): number {
  return Math.round((b.getTime() - a.getTime()) / (1000 * 60 * 60 * 24))
}

export function loadWorkOrders(): WorkOrder[] {
  const rows = readCsvFile(PATHS.workOrders, ';', 'latin1')
  return rows.map((r) => {
    const desc = r.WORK_DESCRIPTION ?? ''
    const action = r.WORK_ORDER_PERFORMED_ACTION ?? ''
    return {
      WO_NO: r.WO_NO ?? '',
      CUSTOMER_NO: Number(r.CUSTOMER_NO),
      CUSTOMER_SITE_NO: Number(r.CUSTOMER_SITE_NO),
      customer_site_name: r.customer_site_name ?? '',
      CONTRACT: r.CONTRACT ?? '',
      WORK_ORDER_TYPE_ENG: r.WORK_ORDER_TYPE_ENG ?? '',
      WORK_TYPE_DESCRIPTION_ENG: r.WORK_TYPE_DESCRIPTION_ENG ?? '',
      WORK_DESCRIPTION: desc,
      WORK_ORDER_PERFORMED_ACTION: action,
      PM_NO: r.PM_NO ?? '',
      PM_ACTION_DESCR: r.PM_ACTION_DESCR ?? '',
      WORK_STARTED_DATETIME: r.WORK_STARTED_DATETIME ?? '',
      WORK_FINISHED_DATETIME: r.WORK_FINISHED_DATETIME ?? '',
      WORKTIME_HOURS: r.WORKTIME_HOURS ?? '',
      WORK_FINISHED_DAYS: r.WORK_FINISHED_DAYS ?? '',
      started: parseFiDate(r.WORK_STARTED_DATETIME),
      finished: parseFiDate(r.WORK_FINISHED_DATETIME),
      hours: Number(r.WORKTIME_HOURS) || 0,
      days: Number(r.WORK_FINISHED_DAYS) || 0,
      equipmentCodes: extractEquipmentCodes(`${desc} ${action}`),
    }
  })
}

export function loadAlarms(): Alarm[] {
  const rows = readCsvFile(PATHS.alarms, ',', 'utf-8')
  return rows.map((r) => ({
    ALERT_ID: r.ALERT_ID ?? '',
    ALERT_EVENT_ID: r.ALERT_EVENT_ID ?? '',
    CUSTOMER_NO: Number(r.CUSTOMER_NO),
    CUSTOMER_SITE_NO: Number(r.CUSTOMER_SITE_NO),
    CUSTOMER_SITE_NAME: r.CUSTOMER_SITE_NAME ?? '',
    EVENT_TIME: parseIsoDate(r.EVENT_TIME),
    EVENT_DESCRIPTION: r.EVENT_DESCRIPTION ?? '',
    ALERT_TYPE: r.ALERT_TYPE ?? '',
    LOG_ONLY: r.LOG_ONLY ?? '',
    WORKORDER_NO: r.WORKORDER_NO ?? '',
    isActionable: !r.LOG_ONLY && r.ALERT_TYPE !== 'INFO',
    equipmentCodes: extractEquipmentCodes(r.EVENT_DESCRIPTION),
  }))
}

export function loadSchedule(): MaintenancePlan[] {
  const rows = readCsvFile(PATHS.schedule, ',', 'utf-8')
  const byPm = new Map<number, MaintenancePlan>()

  for (const r of rows) {
    const pmNo = Number(r.PM_NO)
    const siteNo = Number(r.CUSTOMER_SITE_NO)
    if (!TARGET_SITES.has(siteNo)) continue

    const plan: MaintenancePlan = {
      PM_NO: pmNo,
      CUSTOMER_SITE_NO: siteNo,
      customer_site_name: r.customer_site_name ?? '',
      ACTION_DESCR: r.ACTION_DESCR ?? '',
      ACTION_DESCR_ENG: r.ACTION_DESCR_ENG ?? '',
      INTERVAL: Number(r.INTERVAL) || 0,
      PM_INTERVAL_UNIT: r.PM_INTERVAL_UNIT ?? '',
      PLAN_HRS: Number(r.PLAN_HRS) || 0,
      START_DATE: parseIsoDate(r.START_DATE),
      OBJSTATE: r.OBJSTATE ?? '',
      LAST_UPDATED: r.LAST_UPDATED ?? '',
    }

    const existing = byPm.get(pmNo)
    if (!existing || (r.LAST_UPDATED ?? '') >= existing.LAST_UPDATED) {
      byPm.set(pmNo, plan)
    }
  }

  return [...byPm.values()]
}

export function filterTargetSites<T extends { CUSTOMER_SITE_NO: number }>(rows: T[]): T[] {
  return rows.filter((r) => TARGET_SITES.has(r.CUSTOMER_SITE_NO))
}

export function getReactiveRepairs(wo: WorkOrder[]): WorkOrder[] {
  return wo.filter(
    (r) =>
      r.WORK_ORDER_TYPE_ENG === 'On-demand work' &&
      r.WORK_TYPE_DESCRIPTION_ENG.includes('Repairs') &&
      (r.CONTRACT === 'KH' || r.CONTRACT === 'KT') &&
      r.started &&
      r.started >= ALARM_START_DATE,
  )
}

export function getScheduledWork(wo: WorkOrder[]): WorkOrder[] {
  return wo.filter((r) => r.WORK_ORDER_TYPE_ENG === 'Scheduled work')
}

export function countAlarmsBefore(
  alarms: Alarm[],
  customerNo: number | null,
  siteNo: number,
  end: Date,
  windowDays = 7,
  equipmentCodes?: Set<string>,
): { unique: number; rows: number; first: Date | null } {
  const start = addDays(end, -windowDays)
  let subset = alarms.filter(
    (a) =>
      (customerNo === null || a.CUSTOMER_NO === customerNo) &&
      a.CUSTOMER_SITE_NO === siteNo &&
      a.isActionable &&
      a.EVENT_TIME &&
      a.EVENT_TIME >= start &&
      a.EVENT_TIME < end,
  )

  if (equipmentCodes && equipmentCodes.size > 0) {
    subset = subset.filter((a) => [...a.equipmentCodes].some((c) => equipmentCodes.has(c)))
  }

  if (subset.length === 0) return { unique: 0, rows: 0, first: null }

  const ids = new Set(subset.map((a) => a.ALERT_ID))
  const first = subset.reduce(
    (min, a) => (a.EVENT_TIME && (!min || a.EVENT_TIME < min) ? a.EVENT_TIME : min),
    null as Date | null,
  )
  return { unique: ids.size, rows: subset.length, first }
}

export function intervalToDays(interval: number, unit: string): number | null {
  if (!interval || !unit) return null
  const u = unit.trim().toLowerCase()
  if (u === 'day') return interval
  if (u === 'week') return interval * 7
  if (u === 'month') return interval * 30
  if (u === 'year') return interval * 365
  return null
}

export function computeNextInspection(
  pmNo: number,
  schedule: MaintenancePlan[],
  ehWork: WorkOrder[],
  referenceDate = new Date(),
): Date | null {
  const plan = schedule.find((p) => p.PM_NO === pmNo)
  if (!plan || plan.OBJSTATE === 'Obsolete') return null

  const intervalDays = intervalToDays(plan.INTERVAL, plan.PM_INTERVAL_UNIT)
  if (!intervalDays) return null

  const pmEh = ehWork.filter((w) => w.PM_NO === String(pmNo) && w.finished)
  if (pmEh.length > 0) {
    const last = pmEh.reduce(
      (max, w) => (w.finished! > max ? w.finished! : max),
      pmEh[0].finished!,
    )
    return addDays(last, intervalDays)
  }

  if (!plan.START_DATE) return null
  let next = plan.START_DATE
  while (next <= referenceDate) {
    next = addDays(next, intervalDays)
  }
  return next
}

export type LoadedData = {
  workOrders: WorkOrder[]
  alarms: Alarm[]
  schedule: MaintenancePlan[]
  reactive: WorkOrder[]
  scheduled: WorkOrder[]
}

let cache: LoadedData | null = null

export function loadAllData(): LoadedData {
  if (cache) return cache

  const workOrders = filterTargetSites(loadWorkOrders())
  const alarms = filterTargetSites(loadAlarms())
  const schedule = loadSchedule()

  cache = {
    workOrders,
    alarms,
    schedule,
    reactive: getReactiveRepairs(workOrders),
    scheduled: getScheduledWork(workOrders),
  }
  return cache
}

File: analysis/data_loader.py
"""Shared ETL: load and normalize the three hackathon datasets."""

from __future__ import annotations

import re
from datetime import timedelta

import pandas as pd

from config import (
    ALARMS_PATH,
    EQUIPMENT_CODE_PATTERN,
    SCHEDULE_PATH,
    TARGET_SITES,
    WORK_ORDERS_PATH,
)

_EQUIPMENT_RE = re.compile(EQUIPMENT_CODE_PATTERN, re.IGNORECASE)


def extract_equipment_codes(text: str | float | None) -> set[str]:
    if text is None or (isinstance(text, float) and pd.isna(text)):
        return set()
    return {m.upper().split(".")[0] for m in _EQUIPMENT_RE.findall(str(text))}


def parse_fi_datetime(series: pd.Series) -> pd.Series:
    return pd.to_datetime(series, format="%d.%m.%Y %H:%M", errors="coerce")


def load_work_orders() -> pd.DataFrame:
    df = pd.read_csv(WORK_ORDERS_PATH, sep=";", encoding="latin-1", low_memory=False)
    df["started"] = parse_fi_datetime(df["WORK_STARTED_DATETIME"])
    df["finished"] = parse_fi_datetime(df["WORK_FINISHED_DATETIME"])
    df["hours"] = pd.to_numeric(df["WORKTIME_HOURS"], errors="coerce").fillna(0)
    df["days"] = pd.to_numeric(df["WORK_FINISHED_DAYS"], errors="coerce").fillna(0)
    df["CUSTOMER_SITE_NO"] = pd.to_numeric(df["CUSTOMER_SITE_NO"], errors="coerce")
    df["CUSTOMER_NO"] = pd.to_numeric(df["CUSTOMER_NO"], errors="coerce")
    df["equipment_codes"] = (
        df["WORK_DESCRIPTION"].fillna("").astype(str)
        + " "
        + df.get("WORK_ORDER_PERFORMED_ACTION", pd.Series("", index=df.index)).fillna("").astype(str)
    ).apply(extract_equipment_codes)
    return df


def load_alarms() -> pd.DataFrame:
    df = pd.read_csv(ALARMS_PATH, encoding="utf-8-sig", low_memory=False)
    df["EVENT_TIME"] = pd.to_datetime(df["EVENT_TIME"], errors="coerce")
    df["CUSTOMER_SITE_NO"] = pd.to_numeric(df["CUSTOMER_SITE_NO"], errors="coerce")
    df["CUSTOMER_NO"] = pd.to_numeric(df["CUSTOMER_NO"], errors="coerce")
    df["equipment_codes"] = df["EVENT_DESCRIPTION"].apply(extract_equipment_codes)
    df["is_actionable"] = df["LOG_ONLY"].isna() & (df["ALERT_TYPE"] != "INFO")
    return df


def load_schedule() -> pd.DataFrame:
    df = pd.read_csv(SCHEDULE_PATH, encoding="utf-8-sig", low_memory=False)
    df["CUSTOMER_SITE_NO"] = pd.to_numeric(df["CUSTOMER_SITE_NO"], errors="coerce")
    df["PM_NO"] = pd.to_numeric(df["PM_NO"], errors="coerce")
    df["INTERVAL"] = pd.to_numeric(df["INTERVAL"], errors="coerce")
    df["PLAN_HRS"] = pd.to_numeric(df["PLAN_HRS"], errors="coerce").fillna(0)
    df["START_DATE"] = pd.to_datetime(df["START_DATE"], errors="coerce")
    # Deduplicate: same PM_NO appears ~3x in export
    df = df.sort_values("LAST_UPDATED").drop_duplicates(subset=["PM_NO"], keep="last")
    df = df[df["CUSTOMER_SITE_NO"].isin(TARGET_SITES)]
    return df


def filter_target_sites(df: pd.DataFrame) -> pd.DataFrame:
    return df[df["CUSTOMER_SITE_NO"].isin(TARGET_SITES)].copy()


def get_reactive_repairs(wo: pd.DataFrame, since: str = "2025-01-20") -> pd.DataFrame:
    mask = (
        (wo["WORK_ORDER_TYPE_ENG"] == "On-demand work")
        & (wo["WORK_TYPE_DESCRIPTION_ENG"].str.contains("Repairs", na=False))
        & (wo["CONTRACT"].isin(["KH", "KT"]))
        & (wo["started"] >= pd.Timestamp(since))
    )
    return wo.loc[mask].copy()


def get_scheduled_work(wo: pd.DataFrame) -> pd.DataFrame:
    return wo.loc[wo["WORK_ORDER_TYPE_ENG"] == "Scheduled work"].copy()


def count_alarms_before(
    alarms: pd.DataFrame,
    customer_no: int,
    site_no: int,
    end: pd.Timestamp,
    window_days: int = 7,
    equipment_codes: set[str] | None = None,
    actionable_only: bool = True,
) -> tuple[int, int, pd.Timestamp | None]:
    """Return (unique_alert_count, row_count, first_alarm_time)."""
    start = end - timedelta(days=window_days)
    mask = (
        (alarms["CUSTOMER_NO"] == customer_no)
        & (alarms["CUSTOMER_SITE_NO"] == site_no)
        & (alarms["EVENT_TIME"] >= start)
        & (alarms["EVENT_TIME"] < end)
    )
    if actionable_only:
        mask &= alarms["is_actionable"]
    subset = alarms.loc[mask]
    if equipment_codes:
        subset = subset[subset["equipment_codes"].apply(lambda c: bool(c & equipment_codes))]
    if subset.empty:
        return 0, 0, None
    return subset["ALERT_ID"].nunique(), len(subset), subset["EVENT_TIME"].min()


def interval_to_timedelta(interval: float, unit: str) -> timedelta | None:
    if pd.isna(interval) or not unit:
        return None
    unit = unit.strip().lower()
    n = int(interval)
    if unit == "day":
        return timedelta(days=n)
    if unit == "week":
        return timedelta(weeks=n)
    if unit == "month":
        return timedelta(days=n * 30)
    if unit == "year":
        return timedelta(days=n * 365)
    return None


def compute_next_inspection(
    pm_no: int,
    schedule: pd.DataFrame,
    eh_work: pd.DataFrame,
    reference_date: pd.Timestamp | None = None,
) -> pd.Timestamp | None:
    """Compute next planned inspection date for a PM_NO."""
    ref = reference_date or pd.Timestamp.now()
    plan_rows = schedule.loc[schedule["PM_NO"] == pm_no]
    if plan_rows.empty:
        return None
    plan = plan_rows.iloc[0]
    if plan.get("OBJSTATE") == "Obsolete":
        return None
    delta = interval_to_timedelta(plan["INTERVAL"], plan.get("PM_INTERVAL_UNIT", ""))
    if delta is None:
        return None

    pm_eh = eh_work.loc[eh_work["PM_NO"].astype(str) == str(pm_no)]
    if not pm_eh.empty and pm_eh["finished"].notna().any():
        last = pm_eh["finished"].max()
        return last + delta

    start = plan["START_DATE"]
    if pd.isna(start):
        return None
    next_date = start
    while next_date <= ref:
        next_date = next_date + delta
    return next_date
